File: src/minecraft/net/machinemuse/powersuits/item/ModularItemBase.scala
package net.machinemuse.powersuits.item

import net.minecraft.item.{ItemStack, Item}
import cpw.mods.fml.relauncher.{SideOnly, Side}
import net.machinemuse.general.geometry.Colour
import net.machinemuse.utils.{MuseItemUtils, MuseStringUtils, MuseCommonStrings}
import scala.Predef.String
import net.minecraft.entity.player.EntityPlayer
import net.machinemuse.powersuits.powermodule.misc.{CosmeticGlowModule, TintModule}
import net.machinemuse.utils.MuseMathUtils._
import net.machinemuse.api.electricity.MuseElectricItem
import net.machinemuse.api.{ModuleManager, IModularItem}

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 7:49 PM, 4/23/13
 */
trait ModularItemBase extends Item with IModularItem with MuseElectricItem {
  @SideOnly(Side.CLIENT) override def getColorFromItemStack(stack: ItemStack, par2: Int): Int = {
    val c: Colour = getColorFromItemStack(stack)
    return c.getInt
  }

  def getGlowFromItemStack(stack: ItemStack): Colour = {
    if (!MuseItemUtils.itemHasActiveModule(stack, CosmeticGlowModule.MODULE_GLOW)) {
      return Colour.LIGHTBLUE
    }
    val computedred: Double = ModuleManager.computeModularProperty(stack, CosmeticGlowModule.RED_GLOW)
    val computedgreen: Double = ModuleManager.computeModularProperty(stack, CosmeticGlowModule.GREEN_GLOW)
    val computedblue: Double = ModuleManager.computeModularProperty(stack, CosmeticGlowModule.BLUE_GLOW)
    val colour: Colour = new Colour(clampDouble(computedred, 0, 1), clampDouble(computedgreen, 0, 1), clampDouble(computedblue, 0, 1), 0.8)
    return colour
  }

  def getColorFromItemStack(stack: ItemStack): Colour = {
    if (!MuseItemUtils.itemHasActiveModule(stack, TintModule.MODULE_TINT)) {
      return Colour.WHITE
    }
    val computedred: Double = ModuleManager.computeModularProperty(stack, TintModule.RED_TINT)
    val computedgreen: Double = ModuleManager.computeModularProperty(stack, TintModule.GREEN_TINT)
    val computedblue: Double = ModuleManager.computeModularProperty(stack, TintModule.BLUE_TINT)
    val colour: Colour = new Colour(clampDouble(computedred, 0, 1), clampDouble(computedgreen, 0, 1), clampDouble(computedblue, 0, 1), 1.0F)
    return colour
  }

  @SideOnly(Side.CLIENT) override def requiresMultipleRenderPasses: Boolean = {
    return false
  }

  /**
   * Adds information to the item's tooltip when 'getting' it.
   *
   * @param stack            The itemstack to get the tooltip for
   * @param player           The player (client) viewing the tooltip
   * @param currentTipList   A list of strings containing the existing tooltip. When
   *                         passed, it will just contain the name of the item;
   *                         enchantments and lore are
   *                         appended afterwards.
   * @param advancedToolTips Whether or not the player has 'advanced tooltips' turned on in
   *                         their settings.
   */
  override def addInformation(stack: ItemStack, player: EntityPlayer, currentTipList: java.util.List[_], advancedToolTips: Boolean) {
    MuseCommonStrings.addInformation(stack, player, currentTipList, advancedToolTips)
  }

  def formatInfo(string: String, value: Double): String = {
    return string + '\t' + MuseStringUtils.formatNumberShort(value)
  }

  def getLongInfo(player: EntityPlayer, stack: ItemStack): java.util.List[String] = {
    val info = new java.util.ArrayList[String]
    info.add("Detailed Summary")
    info.add(formatInfo("Armor", getArmorDouble(player, stack)))
    info.add(formatInfo("Energy Storage", getCurrentEnergy(stack)) + 'J')
    info.add(formatInfo("Weight", MuseCommonStrings.getTotalWeight(stack)) + 'g')
    info
  }

  def getArmorDouble(player: EntityPlayer, stack: ItemStack): Double = 0
}

File: src/main/java/net/machinemuse/powersuits/common/items/old/ModularItemBase.java
package net.machinemuse.powersuits.common.items.old;

import net.machinemuse.powersuits.api.electricity.MuseElectricItem;
import net.machinemuse.numina.api.item.IModularItemBase;
import net.machinemuse.numina.geometry.EnumColour;
import net.machinemuse.numina.utils.energy.ElectricItemUtils;
import net.machinemuse.numina.utils.string.MuseCommonStrings;
import net.machinemuse.numina.utils.item.MuseItemUtils;
import net.machinemuse.numina.utils.string.MuseStringUtils;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.ArrayList;
import java.util.List;

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 7:49 PM, 4/23/13
 *
 * Ported to Java by lehjr on 11/4/16.
 */
public class ModularItemBase extends Item implements IModularItemBase {
    private static ModularItemBase INSTANCE;

    public static ModularItemBase getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new ModularItemBase();
        }
        return INSTANCE;
    }

    @SideOnly(Side.CLIENT)
    @Override
    public int getColorFromItemStack(ItemStack stack, int par2) {
        return getColorFromItemStack(stack).getColour().getInt();
    }

    @Override
    public EnumColour getColorFromItemStack(ItemStack stack) {
        return MuseItemUtils.getStackSingleColour(stack);
    }

    @Override
    public void addInformation(ItemStack stack, EntityPlayer player, List currentTipList, boolean advancedToolTips) {
        MuseCommonStrings.addInformation(stack, player, currentTipList, advancedToolTips);
    }

    @Override
    public String formatInfo(String string, double value) {
        return string + '\t' + MuseStringUtils.formatNumberShort(value);
    }

    /* IModularItem ------------------------------------------------------------------------------- */
    @Override
    public List<String> getLongInfo(EntityPlayer player, ItemStack stack) {
        List<String> info = new ArrayList<>();

        info.add("Detailed Summary");
        info.add(formatInfo("Armor", getArmorDouble(player, stack)));
        info.add(formatInfo("Energy Storage", getCurrentMPSEnergy(stack)) + 'J');
        info.add(formatInfo("Weight", MuseCommonStrings.getTotalWeight(stack)) + 'g');
        return info;
    }

    @Override
    public double getArmorDouble(EntityPlayer player, ItemStack stack) {
        return 0;
    }

    @Override
    public double getPlayerEnergy(EntityPlayer player) {
        return ElectricItemUtils.getPlayerEnergy(player);
    }

    @Override
    public void drainPlayerEnergy(EntityPlayer player, double drainEnergy) {
        ElectricItemUtils.drainPlayerEnergy(player, drainEnergy);
    }

    @Override
    public void givePlayerEnergy(EntityPlayer player, double joulesToGive) {
        ElectricItemUtils.givePlayerEnergy(player, joulesToGive);
    }


    /* MuseElectricItem --------------------------------------------------------------------------- */
    @Override
    public double getCurrentMPSEnergy(ItemStack stack) {
        return MuseElectricItem.getInstance().getCurrentMPSEnergy(stack);
    }

    @Override
    public double getMaxMPSEnergy(ItemStack stack) {
        return MuseElectricItem.getInstance().getCurrentMPSEnergy(stack);
    }

    @Override
    public void setCurrentMPSEnergy(ItemStack stack, double energy) {
        MuseElectricItem.getInstance().setCurrentMPSEnergy(stack, energy);
    }

    @Override
    public double drainMPSEnergyFrom(ItemStack stack, double requested) {
        return MuseElectricItem.getInstance().drainMPSEnergyFrom(stack, requested);
    }

    @Override
    public double giveMPSEnergyTo(ItemStack stack, double provided) {
        return MuseElectricItem.getInstance().giveMPSEnergyTo(stack, provided);
    }


    /* Thermal Expansion -------------------------------------------------------------------------- */
    @Override
    public int receiveEnergy(ItemStack stack, int energy, boolean simulate) {
        return MuseElectricItem.getInstance().receiveEnergy(stack, energy, simulate);
    }

    @Override
    public int extractEnergy(ItemStack stack, int energy, boolean simulate) {
        return MuseElectricItem.getInstance().extractEnergy(stack, energy, simulate);
    }

    @Override
    public int getEnergyStored(ItemStack theItem) {
        return MuseElectricItem.getInstance().getEnergyStored(theItem);
    }

    @Override
    public int getMaxEnergyStored(ItemStack theItem) {
        return MuseElectricItem.getInstance().getMaxEnergyStored(theItem);
    }

    /* Mekanism ----------------------------------------------------------------------------------- */
    @Override
    public double getEnergy(ItemStack itemStack) {
        return MuseElectricItem.getInstance().getEnergy(itemStack);
    }

    @Override
    public void setEnergy(ItemStack itemStack, double v) {
        MuseElectricItem.getInstance().setEnergy(itemStack, v);
    }

    @Override
    public double getMaxEnergy(ItemStack itemStack) {
        return MuseElectricItem.getInstance().getMaxEnergy(itemStack);
    }

    @Override
    public double getMaxTransfer(ItemStack itemStack) {
        return MuseElectricItem.getInstance().getMaxTransfer(itemStack);
    }

    @Override
    public boolean canReceive(ItemStack itemStack) {
        return MuseElectricItem.getInstance().canReceive(itemStack);
    }

    @Override
    public boolean canSend(ItemStack itemStack) {
        return MuseElectricItem.getInstance().canSend(itemStack);
    }
}